src/binary-writer.cc

```diff
@@ -739,9 +917,13 @@ void BinaryWriter::WriteExpr(const Func* func, const Expr* expr) {
     case ExprType::Block:
       WriteOpcode(stream_, Opcode::Block);
       WriteBlockDecl(cast<BlockExpr>(expr)->block.decl);
-      WriteExprList(func, cast<BlockExpr>(expr)->block.exprs);
-      WriteOpcode(stream_, Opcode::End);
-      break;
+      if (!cast<BlockExpr>(expr)->block.exprs.empty()) {
+        return BlockState<ExprType::Block>{cast<BlockExpr>(expr)};
+      } else {
+        WriteOpcode(stream_, Opcode::End);
+        break;
+      }
```

Comment:
I think it's a nice shortcut to avoid going through all the stack machinery in trivial cases. I will check if it's okay to remove this duplication for else branches though, because there it gets kinda ugly.